Libraries/MLXLLM/Lora+Data.swift

```
@@ -39,23 +40,111 @@ public func loadLoRAData(url: URL) throws -> [String] {
 }
 
 func loadJSONL(url: URL) throws -> [String] {
-
-    struct Line: Codable {
-        let text: String?
+    let content = try String(contentsOf: url).components(separatedBy: .newlines)
+    var results = [String]()
+    var decoder: ((Data) throws -> String)?
+    
+    for (index, line) in content.enumerated() {
+        guard !line.trimmingCharacters(in: .whitespaces).isEmpty else {
+            continue
+        }
+        
+        if decoder == nil {
+            decoder = try determineDecoder(line: line)
+        }
+        
+        if let decodeFunction = decoder {
+            let data = Data(line.utf8)
+            let decodedLine = try decodeFunction(data)
+            results.append(decodedLine)
+        } else {
+            throw LoRADataError.invalidJSON(line, index + 1)
+        }
     }
+    
+    return results
+}
 
-    return try String(contentsOf: url)
-        .components(separatedBy: .newlines)
-        .filter {
-            $0.first == "{"
+private func determineDecoder(line: String) throws -> (Data) throws -> String {
+    let data = Data(line.utf8)
+    let decoder = JSONDecoder()
+
+    if let _ = try? decoder.decode(ChatStructure.self, from: data) {
+        return { data in
+            let chatStruct = try JSONDecoder().decode(ChatStructure.self, from: data)
+            return chatStruct.messages.map { "\($0.role): \($0.content)" }.joined(separator: "\n")
         }
-        .compactMap {
-            try JSONDecoder().decode(Line.self, from: $0.data(using: .utf8)!).text
+    }
+    
+    if let _ = try? decoder.decode(ToolStructure.self, from: data) {
+        return { _ in line }
```

this is not supported?  Should it produce an error rather than silent failure?